KK_ButtPlugin/ButtplugController.cs

```diff
@@ -102,22 +157,37 @@ IEnumerator RunVibrate()
         {
             while (!flags.isHSceneEnd)
             {
-                if (flags.nowAnimStateName.Equals("OLoop"))
-                {
-                    DoVibrate(1.0f);
-                }
-                else if (!supportedModes.Contains(flags.mode) || !supportedAnimations.Contains(flags.nowAnimStateName))
+                if (!supportedModes.Contains(flags.mode) || !supportedAnimations.Contains(flags.nowAnimStateName))
                 {
                     // stops vibration when not being lewd
                     DoVibrate(0.0f);
+                    yield return new WaitForSecondsRealtime(1.0f / (float)ButtPlugin.VibrationFrequency.Value);
+                    continue;
                 }
-                else
+
+                var animator = flags.lstHeroine[0].chaCtrl.animBody;
+
+                var speed = flags.speedCalc;
+                var strength = 1.0f;
+                var minVibration = 0.2f;
+
+                // service mode goes into OLoop once male excitement exceeds its threshold
+                if (IsOrgasm)
                 {
-                    // vibrate based on the intensity of the player
-                    // minimum vibration above 0 exists so you always feel something along with the animation
-                    DoVibrate(Mathf.Lerp(0.2f, 1.0f, flags.speedCalc));
+                    speed = 1.0f;
+                    minVibration = 0.6f;
                 }
-                yield return new WaitForSeconds(.01f);
+
+                if (ButtPlugin.SyncVibrationWithAnimation.Value)
+                {
+                    // Simple sin based intensity amplification based on normalized position in looping animation
+                    var info = animator.GetCurrentAnimatorStateInfo(0);
+                    var depth = (info.m_NormalizedTime - GetPhase()) % 1;

```

Comment:
m_NormalizedTime is undocumented. No idea why they made it public. normalizedTime is the official property name.

Reply:
ah, my bad.  I found the value from using the bepinex runtime inspector instead of relying on the API.  It's been the best way so far for me to learn what values are mapped to behavior and timing in the game, as none of that is really documented on the API source.